sddl/sddlSplitter.go

```diff
@@ -0,0 +1,403 @@
+package sddl
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+func IffInt(condition bool, tVal, fVal int) int {
+	if condition {
+		return tVal
+	}
+	return fVal
+}
+
+// Owner and group SIDs need replacement
+type SDDLString struct {
+	OwnerSID, GroupSID string
+	DACL, SACL         ACLList
+}
+
+type ACLList struct {
+	Flags      string
+	ACLEntries []ACLEntry
+}
+
+// field 5 and field 6 will contain SIDs.
+// field 5 is a lone SID, but field 6 will contain SIDs under SID(.*)
+type ACLEntry struct {
+	Sections []string
+}
+
+func (s *SDDLString) PortableString() string {
+	output := ""
+
+	if s.OwnerSID != "" {
+		tx, err := TranslateSID(s.OwnerSID)
+
+		if err != nil {
+			output += "O:" + s.OwnerSID
+		} else {
+			output += "O:" + tx
+		}
+	}
+
+	if s.GroupSID != "" {
+		tx, err := TranslateSID(s.GroupSID)
+
+		if err != nil {
+			output += "G:" + s.GroupSID
+		} else {
+			output += "G:" + tx
+		}
+	}
+
+	if s.DACL.Flags != "" || len(s.DACL.ACLEntries) != 0 {
+		output += "D:" + s.DACL.PortableString()
+	}
+
+	if s.SACL.Flags != "" || len(s.SACL.ACLEntries) != 0 {
+		output += "S:" + s.SACL.PortableString()
+	}
+
+	return output
+}
+
+var LiteralSIDRegex = regexp.MustCompile(`SID\(.*?\)`)
+var StringRegex = regexp.MustCompile(`("")|(".*?[^\\]")`)
+
+// PortableString returns a SDDL that's been ported from non-descript, well known SID strings (such as DU, DA, etc.)
+// to domain-specific strings. This allows us to not mix up the admins from one domain to another.
+// Azure Files requires that we do this.
+func (a *ACLList) PortableString() string {
+	output := a.Flags
+
+	for _, v := range a.ACLEntries {
+		output += "("
+
+		for k, s := range v.Sections {
+			// Append a ; after the last section
+			if k > 0 {
+				output += ";"
+			}
+
+			if k == 5 {
+				// This section is a lone SID, so we can make a call to windows and translate it.
+				tx, err := TranslateSID(strings.TrimSpace(s))
+
+				if err != nil {
+					output += s
+				} else {
+					output += tx
+				}
+			} else if k == 6 {
+				// This section will potentially have SIDs unless it's not a conditional ACE.
+				// They're identifiable as they're inside a literal SID container. ex "SID(S-1-1-0)"
+
+				workingString := ""
+				lastAddPoint := 0
+				if v.Sections[0] == "XA" || v.Sections[0] == "XD" || v.Sections[0] == "XU" || v.Sections[0] == "ZA" {
+					// We shouldn't do any replacing if we're inside of a string.
+					// In order to handle this, we'll handle it as a list of events that occur.
+
+					stringEntries := StringRegex.FindAllStringIndex(s, -1)
+					sidEntries := LiteralSIDRegex.FindAllStringIndex(s, -1)
+					eventMap := map[int]int{} // 1 = string start, 2 = string end, 3 = SID start, 4 = SID end.
```

Comment:
Don't bother changing it now, but generally best to use constants to represent such ints-with-special-meanings